opendg/timedelta.py

```diff
@@ -0,0 +1,117 @@
+        }
+
+        if self._is_valid_type(interval_type):
+            self.type = interval_type
+        else:
+            raise ValueError(f'Invalid time granularity type: {interval_type}')
+
+        if isinstance(duration, int):
+            self.dur = duration
+        else:
+            raise ValueError(
+                f'TimeDeltaTG duration should be an integer, got {type(duration)}'
+            )
+
+    @property
+    def interval_type(self) -> str:
+        r"""The time granularity type"""
+        return self.type
+
+    def _is_valid_type(self, type: str) -> bool:
+        r"""Check if the specified time granularity type is valid."""
+        return type in self.time_dict
+
+    def __str__(self):
+        return f'time granularity is {self.type} : {self.time_dict[self.type]}'
+
+    def __len__(self) -> int:
+        r"""Returns the number of duration of the specified time granularity."""
+        return self.dur
+
+    def _convertFromStr(self, interval_type: str, duration: int) -> float:
+        r"""Convert the time granularity to the specified time granularity type."""
+        new_secs = duration * self.time_constant[interval_type]
+        cur_secs = self.get_seconds()
+        return cur_secs / new_secs
+
+    def _convertFromDelta(self, td: 'TimeDeltaTG') -> float:
+        r"""Convert the time granularity to the specified time granularity type."""
+        new_secs = td.get_seconds()
+        cur_secs = self.get_seconds()
+        return cur_secs / new_secs
+
+    def get_seconds(self) -> float:
+        r"""Returns the time granularity in seconds."""
+        return self.dur * self.time_constant[self.type]
+
+    def convert(self, time_delta: Union[str, 'TimeDeltaTG']) -> int:
+        r"""Convert the time granularity to the specified time granularity type, can be either a string or a TimeDeltaTG object.
+
+        Args:
+            time_delta (str or TimeDeltaTG): Type of time granularity. Possible values are 'Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms', 'us', 'ns' or a TimeDeltaTG object
+        """
+        if isinstance(time_delta, str):
+            if time_delta in self.non_convert:
+                raise ValueError(
+                    f'Conversion not allowed for time granularity type {time_delta}'
+                )
+            elif (self._is_valid_type(time_delta)):
+                return self._convertFromStr(time_delta, 1)
+            else:
+                raise ValueError(f'Invalid time granularity type: {time_delta}')
+        elif (type(time_delta) == type(self)):
+            if time_delta.interval_type in self.non_convert:
+                raise ValueError(
+                    f'Conversion not allowed for time granularity type {time_delta.interval_type}'
+                )
+            else:
+                return self._convertFromDelta(time_delta)
+        else:
+            raise ValueError(
+                f'Invalid time granularity type for conversion: {time_delta}'
+            )
```

Comment:
I think this can be somewhat cleaned up:
```suggestion
if isinstance(time_delta, str):
    time_delta = TimeDelta(time_delta)

if not isinstance(time_delta, TimeDelta):
    raise ValueError(f'Invalid type for conversion: {type(time_delta)}')

if type_delta.interval_type in self.non_convert:
    raise ValueError(f'Conversion not allowed for type {time_delta.interval_type}')

return self._convertFromDelta(time_delta)
```

Reply:
agree